Dynamo+LTC: add pybind to set force fallback config and use that in test_extract_compiled_graph.py

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75359
* __->__ #75292


- Follow the convention in [this doc](https://docs.google.com/document/d/1Vi96ITGoK7BW01ZEccexs4pvCQKF4_LdV8w7TfIWPvM/edit) to setup config for ltc force fallback ops. 
- Pybinds are added to read/set the config.
- Use the added pybinds in the unit test which needs to force fallbacks.

Test plan:
```
pytest test/lazy/test_extract_compiled_graph.py
```

Differential Revision: [D35417678](https://our.internmc.facebook.com/intern/diff/D35417678)

shunting314 committed:
Dynamo+LTC: add pybind to set force fallback config and use that in test_extract_compiled_graph.py

[ghstack-poisoned]

shunting314 committed:
Update on "Dynamo+LTC: add pybind to set force fallback config and use that in test_extract_compiled_graph.py"



- Follow the convention in [this doc](https://docs.google.com/document/d/1Vi96ITGoK7BW01ZEccexs4pvCQKF4_LdV8w7TfIWPvM/edit) to setup config for ltc force fallback ops. 
- Pybinds are added to read/set the config.
- Use the added pybinds in the unit test which needs to force fallbacks.

Test plan:
```
pytest test/lazy/test_extract_compiled_graph.py
```

[ghstack-poisoned]

wconstab commented on test/lazy/test_extract_compiled_graph.py:
```diff
@@ -80,19 +82,20 @@ def forward(self, a, b):
+    try:
+        yield None
+    finally:
+        config.set_force_fallback("")
```
maybe better to first do
_orig_force_fallback = config.get_force_fallback()
finally
config.set_force_fallback(_orig_force_fallback)
?

probably unlikely to use multiple values though..

shunting314 replied:
yep, that's better.

shunting314 commented:
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35417678).

wconstab commented on torch/csrc/lazy/core/config.cpp:
```diff
@@ -50,3 +50,21 @@ C10_DEFINE_string(
+namespace torch {
+namespace lazy {
+
+std::string& getLTCForceFallback() {
```
i'm thinking of changing the name to fit in with the `torch_lazy` pattern, but i haven't really thought of what the right name scheme is.  i think we should land this first and then we can consider names later.  my suggestion might be `torch::lazy::get_force_fallback` but i wonder what other people think.  Should 'config' be a namespace, or any other proposal?

shunting314 replied:
Both getLTCForceFallback/getForceFallback and get_force_fallback sounds good as long as we are consistent about the flavor.

wconstab reviewed (approved):
awesome thanks for following up on this @shunting314 💯

shunting314 committed:
Update on "Dynamo+LTC: add pybind to set force fallback config and use that in test_extract_compiled_graph.py"



- Follow the convention in [this doc](https://docs.google.com/document/d/1Vi96ITGoK7BW01ZEccexs4pvCQKF4_LdV8w7TfIWPvM/edit) to setup config for ltc force fallback ops. 
- Pybinds are added to read/set the config.
- Use the added pybinds in the unit test which needs to force fallbacks.

Test plan:
```
pytest test/lazy/test_extract_compiled_graph.py
```

Differential Revision: [D35417678](https://our.internmc.facebook.com/intern/diff/D35417678)

[ghstack-poisoned]

shunting314 commented:
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35417678).

yxd886 commented:
Hello, I am wondering whether there is a way for me to fallback all ops with a flag?

wconstab commented on torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp:
```diff
@@ -107,8 +108,8 @@ static std::unordered_map<std::string, ::torch::lazy::Counter*>
-  static char* force_str = std::getenv("LTC_FORCE_FALLBACK");
-  if (force_str != nullptr) {
+  auto force_str = getLTCForceFallback();
+  if (!force_str.empty()) {
```
@yxd886 this is the code that actually implements the force fallback.

you could modify it to accept a list of ops separated by a comma or something, but currently it only supports a single op, and its both more complex and less efficient to set it up for a long list of many ops.

if you want another mode that just falls back all the time, you could add an env for that and make this function just always return true based on that env.
